File: apps/web/src/lib/blog.ts
export type BlogNodeKind = "folder" | "document";

export type BlogNodeSummary = {
  id: string;
  parent_id: string | null;
  kind: BlogNodeKind;
  name: string;
  slug: string | null;
  is_published: boolean;
  sort_order: number;
  updated_at: string;
};

export type BlogNodeDetail = BlogNodeSummary & {
  content: string;
  created_at: string;
  published_at: string | null;
};

export type BlogTreeNode = BlogNodeSummary & { children: BlogTreeNode[] };

const publicApiBaseUrl =
  process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";
const serverApiBaseUrl =
  typeof window === "undefined"
    ? (process.env.API_INTERNAL_URL ?? publicApiBaseUrl)
    : publicApiBaseUrl;

export function buildBlogTree(nodes: BlogNodeSummary[]): BlogTreeNode[] {
  const byId = new Map<string, BlogTreeNode>();
  for (const node of nodes) byId.set(node.id, { ...node, children: [] });

  const roots: BlogTreeNode[] = [];
  for (const node of byId.values()) {
    const parent = node.parent_id ? byId.get(node.parent_id) : undefined;
    if (parent) parent.children.push(node);
    else roots.push(node);
  }
  const sortTree = (items: BlogTreeNode[]) => {
    items.sort(
      (left, right) =>
        left.sort_order - right.sort_order ||
        (left.kind === right.kind ? left.name.localeCompare(right.name, "zh-CN") : left.kind === "folder" ? -1 : 1),
    );
    for (const item of items) sortTree(item.children);
  };
  sortTree(roots);
  return roots;
}

export async function fetchPublicBlogTree() {
  const response = await fetch(`${serverApiBaseUrl}/api/v1/blog/tree`, {
    cache: "no-store",
  });
  if (!response.ok) return [];
  return (await response.json()) as BlogNodeSummary[];
}

export async function fetchPublishedPost(slug: string) {
  const response = await fetch(
    `${serverApiBaseUrl}/api/v1/blog/posts/${encodeURIComponent(slug)}`,
    { cache: "no-store" },
  );
  if (!response.ok) return null;
  return (await response.json()) as BlogNodeDetail;
}

export const blogApiBaseUrl = publicApiBaseUrl;

File: apps/api/src/app/schemas/blog.py
"""博客树与 Markdown 文档接口模型。"""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

BlogNodeKind = Literal["folder", "document"]


class BlogNodeCreate(BaseModel):
    parent_id: uuid.UUID | None = None
    kind: BlogNodeKind
    name: str = Field(min_length=1, max_length=160)

    @field_validator("name")
    @classmethod
    def trim_name(cls, value: str) -> str:
        return value.strip()


class BlogNodeUpdate(BaseModel):
    parent_id: uuid.UUID | None = None
    name: str | None = Field(default=None, min_length=1, max_length=160)
    slug: str | None = Field(default=None, max_length=180)
    content: str | None = Field(default=None, max_length=2_000_000)
    is_published: bool | None = None
    sort_order: int | None = None

    @field_validator("name", "slug")
    @classmethod
    def trim_optional_text(cls, value: str | None) -> str | None:
        return value.strip() if value is not None else None


class BlogNodeSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    parent_id: uuid.UUID | None
    kind: BlogNodeKind
    name: str
    slug: str | None
    is_published: bool
    sort_order: int
    updated_at: datetime


class BlogNodeDetail(BlogNodeSummary):
    content: str
    created_at: datetime
    published_at: datetime | None
